### PR TITLE
Avoid more warnings with clang and C++20

### PR DESCRIPTION
**Pull request by masterleinad**
Since all of the changed enums are in internal classes any ODR-usage should also be internal. Hence, I didn't provide extra definitions in namescape scope. Also, we don't have them for the corresponding `HIP` variables.

The changes to the comparison operators in `core/src/impl/Kokkos_Atomic_View.hpp` rely on the existence of implicit conversion operators.

### CODE REVIEW

File: core/src/Cuda/Kokkos_Cuda_Instance.hpp
Comment (nliber) [inline core/src/Cuda/Kokkos_Cuda_Instance.hpp:20]: The problem with this change is that under C++14, these are not inline variables. Any code which takes its address, even indirectly when passing it by reference, will break under C++14 (and be magically fixed under C++17) unless there is a corresponding out of line definition for it.

File: core/unit_test/TestAtomics.hpp
Comment (nliber) [inline core/unit_test/TestAtomics.hpp:125]: If we are going to change these anyway, make them hidden friends:

friend bool operator==(const SuperScalar& lhs, const SuperScalar& res) { /*... */ }

Also, while not technically part of this change, is there any reason (performance?) that operator!= doesn't just call operator==?

File: core/src/impl/Kokkos_Atomic_View.hpp
Comment (nliber) [inline core/src/impl/Kokkos_Atomic_View.hpp:304]: While not a comment on this change, the volatile comparisons seem sketchy.

**Review by nliber — CHANGES_REQUESTED** [review]

File: core/unit_test/TestAtomics.hpp
Comment (masterleinad) [reply core/unit_test/TestAtomics.hpp:125]: This is just a test...

File: core/src/Cuda/Kokkos_Cuda_Instance.hpp
Comment (nliber) [reply core/src/Cuda/Kokkos_Cuda_Instance.hpp:20]: Ah, just read the first comment.  Nevermind.

**Review by nliber — APPROVED** [review]

**Comment by masterleinad** [comment]
@ndellingwood should review. Readt from my side. More reviews/opinions welcome. Related to #3716.

**Review by ndellingwood — APPROVED** [review]
lgtm @masterleinad I think intel is now PR tested via the github actions, but let me know if would you like me to do any additional testing

**Comment by ndellingwood** [comment]
I should clarify, these changes were not made to address anything intel-related, but replacing enums with static constexpr variables in the past had caused intel compilation issues, though I don't think I've seen that since intel/15 or intel/16

**Comment by masterleinad** [comment]
> I should clarify, these changes were not made to address anything intel-related, but replacing enums with static constexpr variables in the past had caused intel compilation issues, though I don't think I've seen that since intel/15 or intel/16

It would probably be good to give it a full test run of the Sandia testers. The alternative to the switching to `static constexpr` is to have the `static_cast`s everywhere for now as you do in #3716.

**Comment by crtrott** [comment]
I just merge this and we check tomorrow if it broke anything and then we can revert. I think the likelyhood is very low. We did a similar change in some other places and it passed.

**Pull request merged**
